lib/async_hooks.js

```diff
@@ -0,0 +1,486 @@
+  if (!Number.isSafeInteger(triggerId)) {
+    if (triggerId !== undefined)
+      resource = triggerId;
+    triggerId = initTriggerId();
+  }
+
+  // I'd prefer allowing these checks to not exist, or only throw in a debug
+  // build, in order to improve performance.
+  if (!Number.isSafeInteger(asyncId) || asyncId < 0)
+    throw new RangeError('asyncId must be an unsigned integer');
+  if (typeof type !== 'string' || type.length <= 0)
+    throw new TypeError('type must be a string with length > 0');
+  if (!Number.isSafeInteger(triggerId) || triggerId < 0)
+    throw new RangeError('triggerId must be an unsigned integer');
+
+  for (var i = 0; i < active_hooks_array.length; i++) {
+    if (typeof active_hooks_array[i][init_symbol] === 'function') {
+      runInitCallback(
+        active_hooks_array[i][init_symbol], asyncId, type, triggerId, resource);
+    }
+  }
+
+  // Isn't null if hooks were added/removed while the hooks were running.
+  if (tmp_active_hooks_array !== null) {
+    restoreTmpHooks();
+  }
+}
+
+
+function emitBeforeN(asyncId, triggerId) {
+  for (var i = 0; i < active_hooks_array.length; i++) {
+    if (typeof active_hooks_array[i][before_symbol] === 'function') {
+      runCallback(active_hooks_array[i][before_symbol], asyncId);
+    }
+  }
+
+  if (tmp_active_hooks_array !== null) {
+    restoreTmpHooks();
+  }
+}
+
+
+// Usage: emitBeforeS(asyncId[, triggerId]). If triggerId is omitted then
+// asyncId will be used instead.
+function emitBeforeS(asyncId, triggerId) {
+  // CHECK(Number.isSafeInteger(asyncId) && asyncId > 0)
+  // CHECK(Number.isSafeInteger(triggerId) && triggerId > 0)
+
+  // Validate the ids.
+  if (asyncId < 0 || triggerId < 0) {
+    fatalError('before(): asyncId or triggerId is less than zero ' +
+               `(asyncId: ${asyncId}, triggerId: ${triggerId})`);
+  }
+
+  pushAsyncIds(asyncId, triggerId);
+
+  if (async_hook_fields[kBefore] === 0) {
+    return;
+  }
+
+  emitBeforeN(asyncId, triggerId);
+}
+
+
+// Called from native. The asyncId stack handling is taken care of there before
+// this is called.
+function emitAfterN(asyncId) {
+  if (async_hook_fields[kAfter] > 0) {
+    for (var i = 0; i < active_hooks_array.length; i++) {
+      if (typeof active_hooks_array[i][after_symbol] === 'function') {
+        runCallback(active_hooks_array[i][after_symbol], asyncId);
+      }
+    }
+  }
+
+  if (tmp_active_hooks_array !== null) {
+    restoreTmpHooks();
+  }
+}
+
+
+// TODO(trevnorris): Calling emitBefore/emitAfter from native can't adjust the
+// kIdStackIndex. But what happens if the user doesn't have both before and
+// after callbacks.
+function emitAfterS(asyncId) {
+  emitAfterN(asyncId);
+  popAsyncIds(asyncId);
+}
+
+
+function emitDestroyS(asyncId) {
+  // Return early if there are no destroy callbacks, or on attempt to emit
+  // destroy on the void.
+  if (async_hook_fields[kDestroy] === 0 || asyncId === 0)
```

Comment:
an `asyncId === 0` specifically means the root id (i.e. the point of origin for all asynchronous calls, if their call stack were turned into a tree). Which can never be destroyed, but may be sent during startup. Though I can't remember why I had to check for `asyncId === 0` specifically. Will look into it.